docs/transpile/create-a-transpiler-plugin.ipynb

```diff
@@ -0,0 +1,311 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "58b1f0a7-d62e-4f71-ba01-b8f0beaeeb83",
+   "metadata": {},
+   "source": [
+    "# Create a transpiler plugin\n",
+    "\n",
+    "Creating a [transpiler plugin](transpiler-plugins) is a great way to share your transpilation code with the wider Qiskit community, allowing other users to benefit from the functionality you've developed. Thank you for your interest in contributing to the Qiskit community!\n",
+    "\n",
+    "Before you create a transpiler plugin, you need to decide what kind of plugin is appropriate for your situation. There are three kinds of transpiler plugins:\n",
+    "- [**Transpiler stage plugin**](/api/qiskit/transpiler_plugins). Choose this if you are defining a pass manager that can be substituted for one of the [6 stages](transpiler-stages) of a preset staged pass manager.\n",
+    "- [**Unitary synthesis plugin**](/api/qiskit/qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin). Choose this if your transpilation code takes as input a unitary matrix and outputs a description of a quantum circuit implementing that unitary.\n",
+    "- [**High-level synthesis plugin**](/api/qiskit/qiskit.transpiler.passes.synthesis.plugin.HighLevelSynthesisPlugin). Choose this if your transpilation code takes as input a \"high-level object\" such as a Clifford operator or a linear function and outputs a description of a quantum circuit implementing that high-level object. High-level objects are represented by subclasses of the [Operation](/api/qiskit/qiskit.circuit.Operation) class.\n",
+    "\n",
+    "Once you've determined which kind of plugin to create, follow these steps to create the plugin:\n",
+    "\n",
+    "1. Create a subclass of the appropriate abstract plugin class:\n",
+    "   - [PassManagerStagePlugin](/api/qiskit/qiskit.transpiler.preset_passmanagers.plugin.PassManagerStagePlugin) for a transpiler stage plugin,\n",
+    "   - [UnitarySynthesisPlugin](/api/qiskit/qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin) for a unitary synthesis plugin, and\n",
+    "   - [HighLevelSynthesisPlugin](/api/qiskit/qiskit.transpiler.passes.synthesis.plugin.HighLevelSynthesisPlugin)  for a high-level synthesis plugin.\n",
+    "2. Expose the class as a [setuptools entry point](https://setuptools.pypa.io/en/latest/userguide/entry_point.html) in the package metadata, typically by adding an `entry-points` table in `pyproject.toml`.\n",
+    "\n",
+    "There is no limit to the number of plugins a single package can define, but each plugin must have a unique name. The name `default` is used by Qiskit itself and can't be used in a plugin.\n",
```

Comment:
I might link to the table here https://docs.quantum.ibm.com/api/qiskit/transpiler_plugins#plugin-stages as it has a list of all the reserved plugins names: https://docs.quantum.ibm.com/api/qiskit/transpiler_plugins#plugin-stages

Reply:
The table you link has the reserved name for transpiler stage plugins, right? Are there reserved names for unitary and high-level synthesis plugins?

Another question: I see that `toqm` appears on that list. It is not a built-in plugin though (it's a third-party one). So is it really supposed to be reserved?

Reply:
Oh, that's a good point. There are but I'm not sure they're clearly documented in the table like for the stage plugins. Basically everything in qiskit's `pyproject.toml` should be treated as reserved. We should list that in the docs for the synthesis plugins too.

The toqm one is a error, before we had the plugin interface in qiskit  it was a reserved word because the equivalent of the plugin lived in qiskit and there was an optional dependency on qiskit-toqm. After the introduction of the stage plugin interface qiskit-toqm just added a native plugin.

Reply:
Alright I linked to the table for transpiler stage plugins and also listed out the reserved names for the synthesis plugins. I also opened https://github.com/Qiskit/qiskit/pull/11792 to remove `toqm` from the reserved name list.